### DISCUSSION TITLE
Task Delegate Conversions

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#1598] [Language Ideas]
Hi All,
It would be nice if Funcs and Actions could be automatically converted to their task equivalents.  For example:
````

//This takes a task<bool>
public async Task DoSomething(Func<Task<bool>> Method){
  var ret = await Method();
  if(!ret){
    throw new Exception();
  }
}

public Task<Bool> Test1(){
  return Task.FromResult(true);
}

public bool Test2(){
  return True;
}

//All of these should work:
DoSomething(Test1);
DoSomething(Test2);
DoSomething(()=> await Test1());
DoSomething(() => Test2());
DoSomething(()=> true);

````

The conversion from a func/action to a task would essentially be call the func/action and then return either Task.CompletedTask (if it was an action) or Task.FromResult(...) if it was a func.

**Comment by ufcpp** [comment]
related to LINQ section in https://github.com/dotnet/csharplang/blob/master/proposals/async-streams.md

**Comment by jnm2** [comment]
Not sure I like the idea of hiding that expense and complexity in an automatic conversion.
